src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -182,7 +182,12 @@ void uct_tcp_sockcm_ep_handle_error(uct_tcp_sockcm_ep_t *cep, ucs_status_t statu
                  cep->fd, ucs_status_string(async_status));
     }
 
-    uct_tcp_sockcm_ep_invoke_error_cb(cep, status);
+    /* if the private data pack callback failed, then the upper layer already
+     * knows about it since it failed in it. in this case, no need to invoke
+     * another upper layer callback. */
+    if (!(cep->state & UCT_TCP_SOCKCM_EP_PACK_CB_FAILED)) {
```

Comment:
there is also state check in uct_tcp_sockcm_ep_handle_remote_disconnect
maybe we can unite uct_tcp_sockcm_ep_handle_error and uct_tcp_sockcm_ep_handle_remote_disconnect functions to check the state in one place only

Reply:
what do you mean? those are different places in the flow. 
uct_tcp_sockcm_ep_handle_remote_disconnect is called before uct_tcp_sockcm_ep_handle_error.
uct_tcp_sockcm_ep_handle_error can be called without calling uct_tcp_sockcm_ep_handle_remote_disconnect

Reply:
sorry, i meant: unite uct_tcp_sockcm_ep_handle_error with **uct_tcp_sockcm_ep_handle_event_status**

Reply:
how about i remove the uct_tcp_sockcm_ep_handle_error() function?
it's called only from uct_tcp_sockcm_ep_handle_event_status() anyway. i could put its code in uct_tcp_sockcm_ep_handle_event_status and then all these state checks would be only in that function.

Reply:
yes, that sounds good